core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3766,6 +3771,104 @@ private static void fix(List<RexNode> operands, int before, int after) {
     }
   }
 
+  /**
+   * Determine whether there is mapping between project input and output fields.
+   * Bail out if sort relies on non-trivial expressions.
+   */
+  private static boolean isCollationOnTrivialExpr(
+      List<RexNode> projects, RelDataTypeFactory typeFactory,
+      Mappings.TargetMapping map, RelFieldCollation fc, boolean passDown) {
+    final int index = fc.getFieldIndex();
+    int target = map.getTargetOpt(index);
+    if (target < 0) {
+      return false;
+    }
+
+    final RexNode node = passDown ? projects.get(index) : projects.get(target);
+    if (node.isA(SqlKind.CAST)) {
+      // Check whether it is a monotonic preserving cast
+      final RexCall cast = (RexCall) node;
+      RelFieldCollation newFc = Objects.requireNonNull(RexUtil.apply(map, fc));
+      final RexCallBinding binding =
+          RexCallBinding.create(typeFactory, cast,
+              ImmutableList.of(RelCollations.of(newFc)));
+      if (cast.getOperator().getMonotonicity(binding)
+          == SqlMonotonicity.NOT_MONOTONIC) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  public static Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
```

Comment:
Depends on if we want move all most all the code to RelOptUtil, I feel like only `isCollationOnTrivialExpr` can be moved to RelOptUtil.

At least `RelOptUtil.passThroughTraits` and `RelOptTuil. deriveTraits` seems to me, could encourage people write all `passThroughTraits` and `deriveTraits` into `RelOptUtil`, because names here are too generic.

For example, I might ask, why not move implementations in Enumerable Joins to `RelOptUtil`, etc.

Reply:
I had the same feeling. How about add a package level internal interface for EnumerableProject and EnumerableCalc with default implementation so that both operators just need to implement the interface?

Reply:
@hsyuan Instead of adding an interface, how about adding an EnumTraitUtils to hold these methods?  IMO, an interface seems not necessary since only Calc and Project should implement it.

Reply:
@amaliujia We could change the name to clarify what it is doing.

Reply:
Updated.